[Wiki] Decision support for publishing target locations (#1681)
Azure/ResourceModules: docs/wiki/The CI environment - Publishing.md

> This section provides an overview of the principles the publishing is built upon, how it is set up, and how you can interact with it.

Author: Regarding relocation:
- Move to 'Solution Creation' after 'Orchestration' section

Author: done

  > **Note:** Template Spec names have a maximum of 90 characters

- **Bicep Registry**<p>
  A [Bicep Registry](https://docs.microsoft.com/en-us/azure/azure-resource-manager/bicep/private-module-registry) is an Azure Container Registry that can be used to store & reference Bicep modules.<p>
  For deployment, it is recommended to apply a [template-orchestrated](./Solution%20creation#Orchestration-overview) approach. As Bicep supports the Bicep registry as linked templates, this approach enables you to fully utilize Azure's parallel deployment capabilities.

- **Azure DevOps Universal Packages**<p>
  A [Universal Package](https://docs.microsoft.com/en-us/azure/devops/artifacts/quickstarts/universal-packages) is a packaged folder in an Azure DevOps artifact feed.<p>
  As such, it contains the content of a CARML module 'as-is', including the template file(s), ReadMe file(s) and test file(s). <p>
  For deployment, it is recommended to use Universal Packages only for a [pipeline-orchestrated](./Solution%20creation#Orchestration-overview) approach - i.e., each job would download a single package and deploy it. <p>

>   Technically, it would be possible to also use Universal Packages for the template-orchestrated approach, by downloading all packages into a specific location first, and then reference them. Given the indirect nature of this approach, this is however not recommended.

Author: Add symbol

Author: done

> **Note**<br>
> The `version` used for publishing any artifact is the same for all three target locations, which reduces the maintenance effort.

# Considerations

> While the CARML CI-environment can publish to all locations mentioned above, you may effectively decide to configure only subset as per your requirements. To help you with the decision, the following content provides you with an overview of the possibilities of each target location.

Reviewer: ```suggestion
While the CARML CI environment can publish to all locations mentioned above, you may effectively decide to configure only a subset as per your requirements. To help you with the decision, the following content provides you with an overview of the possibilities of each target location.
```

## Outline
- **Template Specs**<p>

>   A [Template Spec](https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/template-specs?tabs=azure-powershell) is an Azure resource with the purpose of storing & referencing Azure Resource Manager (ARM) templates. <p>

Reviewer: I'd mention the template specs name length limitation

Author: added

  When publishing Bicep modules as Template Specs, the module is compiled - and the resulting ARM template is uploaded as a Template Spec resource version to a Resource Group of your choice.
  For deployment, it is recommended to apply a [template-orchestrated](./Solution%20creation#Orchestration-overview) approach. As Bicep supports the Template-Specs as linked templates, this approach enables you to fully utilize Azure's parallel deployment capabilities.
  > **Note:** Even though the published resource is an ARM template, you can reference it in you Bicep template as a remote module like it would be native Bicep.

- **Bicep Registry**<p>
  A [Bicep Registry](https://docs.microsoft.com/en-us/azure/azure-resource-manager/bicep/private-module-registry) is an Azure Container Registry that can be used to store & reference Bicep modules.<p>
  For deployment, it is recommended to apply a [template-orchestrated](./Solution%20creation#Orchestration-overview) approach. As Bicep supports the Bicep registry as linked templates, this approach enables you to fully utilize Azure's parallel deployment capabilities.

- **Azure DevOps universal packages**<p>

>   A [Universal Package](https://docs.microsoft.com/en-us/azure/devops/artifacts/quickstarts/universal-packages) is a packaged folder in an Azure DevOps artifact feed.<p>

Reviewer: I'd mention the native semver support

Author: Added

Reviewer: Besides the downside of not allowing a custom version, isn't a plus the fact that you can reference latest major/minor/patch using wildcards?

Author: The benefit for the Bicep Registry & Template Specs here are that you are not suprised by a breaking change.

>   As such, they contain the content of a CARML module 'as-is', including the template file(s), ReadMe file(s) and test file(s). <p>

Reviewer: ```suggestion
  As such, it contains the content of a CARML module 'as-is', including the template file(s), ReadMe file(s) and test file(s). <p>
```

  For deployment, it is recommended to use Universal Packages only for a [pipeline-orchestrated](./Solution%20creation#Orchestration-overview) approach - i.e., each job would download a single package and deploy it. <p>
  Technically, it would be possible to also use Universal Packages for the template-orchestrated approach, by downloading all packages into a specific location first, and then reference them. Given the indirect nature of this approach, this is however not recommended.

> 

Reviewer: ```suggestion
```
